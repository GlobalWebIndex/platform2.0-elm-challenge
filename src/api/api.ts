import { FetchCatImagesParams } from '../types/theCatApi';
import requestWithErrorHandling from './requestWithErrorHandling';

const fetchCatImages = (params: FetchCatImagesParams) =>
  requestWithErrorHandling({
    method: 'get',
    url: 'images/search',
    params,
  });

const fetchBreedOptions = () =>
  requestWithErrorHandling({
    method: 'get',
    url: 'breeds',
  });

const fetchCategoryOptions = () =>
  requestWithErrorHandling({
    method: 'get',
    url: 'categories',
  });

const fetchBreedInfo = (id: string) =>
  requestWithErrorHandling({
    method: 'get',
    url: `images/search?breed_ids=${id}&limit=10`,
  });

const fetchAllFavourites = () =>
  requestWithErrorHandling({
    method: 'get',
    url: 'favourites',
  });

const fetchFavouriteCat = (favourite_id: string) =>
  requestWithErrorHandling({
    method: 'get',
    url: `favourites/${favourite_id}`,
  });

const saveImageAsFavourite = (image_id: string) =>
  requestWithErrorHandling({
    method: 'post',
    url: 'favourites',
    data: {
      image_id,
    },
  });

const removeImageFromFavourites = (favourite_id: string) =>
  requestWithErrorHandling({
    method: 'delete',
    url: `favourites/${favourite_id}`,
  });

const fetchCatDetails = (image_id: string) =>
  requestWithErrorHandling({
    method: 'get',
    url: `images/${image_id}`,
  });

export {
  fetchCatDetails,
  fetchCatImages,
  fetchBreedOptions,
  fetchCategoryOptions,
  fetchBreedInfo,
  fetchAllFavourites,
  fetchFavouriteCat,
  saveImageAsFavourite,
  removeImageFromFavourites,
};
