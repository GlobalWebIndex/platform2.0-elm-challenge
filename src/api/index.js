import axios from 'axios';

const API_KEY = process.env.REACT_APP_API_KEY;

class Cats {
  static getCats(params) {
    return axios({
      method: 'get',
      url: 'https://api.thecatapi.com/v1/images/search',
      headers: {
        'Content-Type': 'application/json',
      },
      params,
    });
  }

  static getBreeds() {
    return axios({
      method: 'get',
      url: 'https://api.thecatapi.com/v1/breeds',
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  static getFavourites() {
    return axios({
      method: 'get',
      url: 'https://api.thecatapi.com/v1/favourites',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': API_KEY,
      },
    });
  }

  static saveFavourite(imageId) {
    return axios({
      method: 'post',
      url: 'https://api.thecatapi.com/v1/favourites',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': API_KEY,
      },
      data: {
        image_id: imageId,
        sub_id: 'vissarion',
      },
    });
  }

  static deleteFavourite(favouriteId) {
    return axios({
      method: 'delete',
      url: `https://api.thecatapi.com/v1/favourites/${favouriteId}`,
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': API_KEY,
      },
    });
  }
}

export default Cats;
