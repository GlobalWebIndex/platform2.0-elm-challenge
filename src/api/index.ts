import * as Client from './client';

export { Client };
export * from './Response';
