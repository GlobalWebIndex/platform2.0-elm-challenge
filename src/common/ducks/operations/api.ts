import { put, call } from 'redux-saga/effects';

import { CommonActionCreators } from 'common/ducks';

/**
 * Higher order function to provide common functionality when making
 * API calls. It is called from a saga effect;
 *
 * .i.e
 * ```js
 *  yield takeLatest(ActionNames, composeWithCommons(handler))
 * ```
 *
 * Saga will call the composeWithCommons with the dispatched action
 * ```
 *  composeWithCommons(handler)(action)
 * ```
 *
 * - Dispatches loading actions
 * - Displays notifications
 * - Handles 500 errors
 *
 * @param generator
 * @param configuration - ```{ withLoader: boolean }``` An object for configuring the common behavior of the saga
 */
function composeWithCommons(
  generator: { context: unknown; fn: (this: unknown, ...args: any[]) => any },
  { withLoader = true } = {}
) {
  return function* (action: any) {
    try {
      if (withLoader) {
        yield put(CommonActionCreators.uiLoadingStarted());
      }

      yield call(generator, { action });
    } catch (error: any) {
      if (error.response?.status === 500) {
        yield put(
          CommonActionCreators.setNotificationMessage(
            error.response?.data,
            'error'
          )
        );
      } else {
        yield put(
          CommonActionCreators.setNotificationMessage(
            error.response?.data,
            'error'
          )
        );
      }
    } finally {
      if (withLoader) {
        yield put(CommonActionCreators.uiLoadingFinished());
      }
    }
  };
}

export { composeWithCommons };
