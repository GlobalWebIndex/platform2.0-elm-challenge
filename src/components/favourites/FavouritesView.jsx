import React from "react";
import Header from "../common/Header";
import FavouritesList from "./FavouritesList";

const FavouritesView = () => {
  return (
    <>
      <Header title="Your favourite 🐈 images..." />
      <FavouritesList />
    </>
  );
};

export default FavouritesView;
