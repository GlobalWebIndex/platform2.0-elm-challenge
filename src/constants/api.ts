export const requestSettings = {
  headers: {
    "content-type": "application/json",
    "x-api-key":
      "live_fnyt6VZN133Ttoe8VmXfAJAAWqKOuw9J1EP9gRawpjqX5Z6ijbvQjyqcmSH2SoCP",
    "Access-Control-Allow-Origin": "*",
  },
};
