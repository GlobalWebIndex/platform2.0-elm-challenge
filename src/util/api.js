import axios from "axios";
import {
  Endpoints,
  CatsToFetchLimit,
  ImagesByBreedLimit,
  API_HTTP_METHODS,
} from "../constants";

/**
 * Setting the API client - using axios.create instance
 */
const APIClient = axios.create({
  baseURL: "https://api.thecatapi.com/v1/",
  headers: {
    common: {
      "x-api-key": process.env.REACT_APP_CAT_API_KEY,
    },
  },
});

/**
 * Utility function to perform the API requests.
 *
 * @param {string} source
 *    URL to be requested
 * @param {*} params
 *    Required params
 * @param {*} method
 *    get/post/delete
 * @returns response data
 */
const fetchAPI = async (source, params = {}, method = API_HTTP_METHODS.GET) => {
  const response = await APIClient[method](source, { ...params });
  return response.data || [];
};

/**
 * Fetch random cats based on the page number given as an arg.
 * API link: https://docs.thecatapi.com/api-reference/images/images-search
 * @param {Number} page
 * @returns
 */
export function fetchCatsList(page = 1) {
  return fetchAPI(Endpoints.ImageSearch, {
    params: {
      page,
      limit: CatsToFetchLimit,
      size: "small",
    },
  });
}

/**
 * Returns the available breeds
 * API link: https://docs.thecatapi.com/api-reference/breeds/breeds-list
 * @returns
 */
export function fetchBreeds() {
  return fetchAPI(Endpoints.Breeds);
}

/**
 * Fetch details for a cat
 * API link: https://docs.thecatapi.com/api-reference/images/images-get
 * @param {*} imageId
 * @returns
 */
export function fetchCatDetails(imageId) {
  return fetchAPI(`${Endpoints.CatDetails}${imageId}`);
}

/**
 * Fetch favourites list
 * API link: https://docs.thecatapi.com/api-reference/favourites/favourites-list
 * @returns
 */
export function fetchFavourites() {
  return fetchAPI(Endpoints.Favourites);
}

/**
 * Fetch Cat images by breed
 * API link: https://docs.thecatapi.com/api-reference/images/images-search
 * @returns
 */
export function fetchImagesByBreed(breedId) {
  return fetchAPI(Endpoints.ImageSearch, {
    params: {
      size: "small",
      breed_ids: breedId,
      limit: ImagesByBreedLimit,
    },
  });
}

/**
 * Submit a favourite cat to the API.
 * API link: https://docs.thecatapi.com/api-reference/favourites/favourites-post
 * @param {*} imageId
 * @returns
 */
export function addFavourite(imageId) {
  return fetchAPI(
    Endpoints.Favourites,
    {
      image_id: imageId,
      // Assuming that sub_id is not used for this app YET.
      // Probably we could have different user account ids in the future
      // OR even send a uuid to simulate random users but wouldn't make sense for now.
      // More Info: https://docs.thecatapi.com/sub_id
    },
    API_HTTP_METHODS.POST
  );
}

/**
 * Remove a favourite that was already submitted
 * API link: https://docs.thecatapi.com/api-reference/favourites/favourites-delete
 * @param {*} favouriteId
 * @returns
 */
export async function removeFavourite(favouriteId) {
  return fetchAPI(
    `${Endpoints.Favourites}${favouriteId}`,
    {},
    API_HTTP_METHODS.DELETE
  );
}
