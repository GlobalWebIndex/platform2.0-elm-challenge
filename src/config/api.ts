import axios from "axios"
import { API_PATH, API_KEY } from "./constants"

export const fetchData = async (endpoint:string, config:object) => {
    return axios({
        url: `${API_PATH}${endpoint}`,
        headers: {'x-api-key': API_KEY},
        method: "get",
        responseType: "json",
        ...config
    });
};