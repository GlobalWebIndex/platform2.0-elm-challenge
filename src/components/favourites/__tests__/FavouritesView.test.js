import { render, screen, waitFor } from "@testing-library/react";
import FavouritesView from "../FavouritesView";
import * as api from "../../../util/api.js";
import { BrowserRouter } from "react-router-dom";

describe("FavouritesView", () => {
  it("renders", async () => {
    jest.spyOn(api, "fetchFavourites").mockResolvedValue([
      {
        image: {
          id: "12",
          url: "https://cat.com",
        },
      },
    ]);

    const { container } = render(<FavouritesView />, {
      wrapper: BrowserRouter,
    });

    expect(screen.getByTestId("loading")).toBeInTheDocument();

    await waitFor(() => {
      screen.getByTestId("card-image");
    });

    expect(container).toMatchSnapshot();
  });

  it("renders an empty list", async () => {
    jest.spyOn(api, "fetchFavourites").mockResolvedValue([]);

    render(<FavouritesView />, {
      wrapper: BrowserRouter,
    });

    await waitFor(() => {
      expect(screen.getByTestId("empty-favourite-list")).toBeInTheDocument();
    });
  });
});
