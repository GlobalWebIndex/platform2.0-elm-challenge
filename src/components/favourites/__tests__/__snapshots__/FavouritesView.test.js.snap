// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouritesView renders 1`] = `
<div>
  <header
    class="title"
    data-testid="header"
  >
    <h1>
      Your favourite 🐈 images...
    </h1>
  </header>
  <section
    class="gridContainer-0-2-1"
  >
    <a
      class="cardItem-0-2-2 cardItem-d0-0-2-6"
      href="/cat/12"
    >
      <img
        alt="id 12"
        class="cardImage-0-2-3"
        data-testid="card-image"
        loading="lazy"
        src="https://cat.com"
      />
    </a>
  </section>
</div>
`;
